src/main.rs

```diff
@@ -113,43 +140,60 @@ pub fn main() -> std::io::Result<()> {
     }
 
     let mut input = String::new();
+    let mut broken_links = vec![];
     if !&matches.free.is_empty() {
         for filename in &matches.free {
             let real_path = PathBuf::from(filename);
             let mut f = File::open(&real_path).expect("file not found");
             f.read_to_string(&mut input)
                 .expect("something went wrong reading the file");
             if matches.opt_present("events") {
-                print_events(&input, opts);
+                print_events(&input, opts, &mut broken_links);
             } else if matches.opt_present("dry-run") {
-                dry_run(&input, opts);
+                dry_run(&input, opts, &mut broken_links);
             } else {
-                pulldown_cmark(&input, opts);
+                pulldown_cmark(&input, opts, &mut broken_links);
             }
         }
     } else {
         let _ = io::stdin().lock().read_to_string(&mut input);
         if matches.opt_present("events") {
-            print_events(&input, opts);
+            print_events(&input, opts, &mut broken_links);
         } else if matches.opt_present("dry-run") {
-            dry_run(&input, opts);
+            dry_run(&input, opts, &mut broken_links);
         } else {
-            pulldown_cmark(&input, opts);
+            pulldown_cmark(&input, opts, &mut broken_links);
         }
     }
+
+    if matches.opt_present("reject-broken-links") && !broken_links.is_empty() {
+        eprintln!("Error: {} broken links:", broken_links.len());
+        for link in broken_links {
+            let start = link.span.start;
+            let end = link.span.end;
+            let reference = link.reference;
+            eprintln!("[{start}-{end}]: {reference}");
+        }
+        std::process::exit(1);
+    }
+
     Ok(())
 }
 
-pub fn pulldown_cmark(input: &str, opts: Options) {
-    let mut p = Parser::new_ext(input, opts);
+pub fn pulldown_cmark<'i>(input: &str, opts: Options, broken_links: &mut Vec<BrokenLink<'static>>) {
+    let mut p = Parser::new_with_broken_link_callback(
+        input,
+        opts,
+        Some(|link: BrokenLink<'_>| {
```

Comment:
This closure is written exactly equal three times, violating the DRY principle. Please try to fix this code duplication.

Reply:
Each closure is a single instruction where values are piped as-is, with the only non-trivial transformation being the `reference.into_string().into()` part to clone the string.

I don't think factoring out the closure into a function would reduce semantic duplication, and it would make the code significantly more complex lifetime-wise.

Reply:
I guess I could add a `to_static` method to BrokenLink, which would reduce each closure to a single line, which would be less duplicate-y.

Reply:
Done.

Reply:
Yeah, it looks cleaner now! But cannot the final `None` lines be removed?

Reply:
Short version: not without making other code hard to read.

Reply:
OK, then ready to be merged. Thanks!